### PR TITLE
sidecar-injection: provide granular control for injection

### LINKED ISSUE #1617: Decouple automatic sidecar injection from namespace monitoring
**Please describe the Improvement and/or Feature Request**
Currently, when a namespace is labeled with `openservicemesh.io/monitored-by` label, the namespace is enabled for automatic sidecar injection.

Based on previous discussions, it was decided that it would be better to decouple automatic sidecar injection from the namespace monitoring. The use case is that to selectively onboard pods in a namespace to be injected with sidecars, automatically injecting a sidecar on all pods within the namespace is not desirable. A user could configure a namespace for monitoring, and selectively annotate pods for sidecar injection.

**Scope (please mark with X where applicable)**
- New Functionality          [X]
- Install                    [ ]
- SMI Traffic Access Policy  [ ]
- SMI Traffic Specs Policy   [ ]
- SMI Traffic Split Policy   [ ]
- Permissive Traffic Policy  [ ]
- Ingress                    [ ]
- Egress                     [ ]
- Envoy Control Plane        [ ]
- CLI Tool                   [ ]
- Metrics                    [ ]
- Certificate Management     [ ]
- Sidecar Injection          [ ]
- Logging                    [ ]
- Debugging                  [ ]
- Tests                      [ ]
- CI System                  [ ]
- Project Release            [ ]

**Possible use cases**
- Allow per pod sidecar injection
- Allow per namespace sidecar injection

### PR DESCRIPTION
**Pull request by shashankram**
Currently it is not possible to enroll specific pods within
the mesh for sidecar injection without enrolling the entire
namespace. This behavior is not desirable when several
application pods run in the same namespace and only certain
pods need to be enrolled in the mesh with a sidecar.

This change decouples the sidecar injection for pods from
the namespace monitoring workflow.

A pod will be injected with the sidecar if its namespace
is monitored, and either of the following are true:

1. The pod is explicitly enabled for injection
2. The namespace is enabled for injection and the pod
   is not explicitly disabled for injection

This change also extends the `osm namespace add` command
with an `--enable-sidecar-injection` flag which automatically
adds the sidecar injection annotation on the namespace.
Similarly, the `osm namespace remove` command is updated
to remove the annotation if present.

It also fixes an existing bug in the `osm namespace remove` command
where all the labels are removed instead of just the monitor label.

Resolves #1617

Signed-off-by: Shashank Ram <shashank08@gmail.com>

- New Functionality      [ ]
- Documentation          [X]
- Install                [X]
- Control Plane          [ ]
- CLI Tool               [X]
- Certificate Management [ ]
- Networking             [ ]
- Metrics                [ ]
- SMI Policy             [ ]
- Security               [ ]
- Tests / CI System      [ ]
- Other                  [ ]

Please answer the following questions with yes/no.

- Does this change contain code from or inspired by another project? If so, did you notify the maintainers and provide attribution?
`No`

### CODE REVIEW

File: docs/patterns/sidecar_injection.md [hunk 1]
```diff
@@ -4,6 +4,54 @@ Services participating in the service mesh communicate via sidecar proxies insta
...
-Since sidecars are automatically injected to pods deployed in OSM monitored namespaces, pods that should not be a part of the service mesh but belong to monitored namespaces need to be explicitly annotated to disable automatic sidecar injection. Using the annotation `"openservicemesh.io/sidecar-injection": "disabled"` on the POD will inform OSM to not inject the sidecar on the POD.
\ No newline at end of file
+### Enabling Automatic Sidecar Injection
+
+Automatic Sidecar injection can be enabled in the following ways:
+
+- While enrolling a namespace into the mesh using `osm` cli: `osm namespace add <namespace> --enable-sidecar-injection`
+
+- Using `kubectl` to annotate individual namespaces and pods to enable sidecar injection:
+
```
Comment (michelleN) [inline docs/patterns/sidecar_injection.md:16]: Should we note here that label `openservicemesh.io/monitored-by=<mesh-name>` is also required on a namespace for automatic sidecar injection?

File: docs/patterns/sidecar_injection.md [hunk 2]
```diff
@@ -4,6 +4,54 @@ Services participating in the service mesh communicate via sidecar proxies insta
...
+  metadata:
+    name: test
+    annotations:
+      'openservicemesh.io/sidecar-injection': 'enabled'
+  ```
+
+  Pods will be injected with a sidecar only if the following conditions are met:
+  1. The namespace to which the pod belongs is a monitored namespace.
+  2. The pod is explicitly enabled for the sidecar injection, OR the namespace to which the pod belongs is enabled for the sidecar injection and the pod is not explicitly disabled for sidecar injection.
+
```
Comment (michelleN) [inline docs/patterns/sidecar_injection.md:38]: What is the behavior if this monitor label doesn't exist but the sidecar injection annotation does exist on a namespace?

Comment (shashankram) [reply docs/patterns/sidecar_injection.md:38]: The monitor label is a prerequisite for OSM to monitor any resources in a namespace, including sidecar injection.

File: pkg/injector/webhook_test.go [hunk 3]
```diff
@@ -93,3 +97,288 @@ func (mc mockCertificate) GetCertificateChain() []byte           { return []byte
...
+			Expect(enabled).To(BeFalse())
+			Expect(err).To(BeNil())
+		})
+	})
+
+	Context("when the inject annotation does not exist", func() {
+		It("should return false to indicate the annotation does not exist", func() {
+			annotation := map[string]string{}
+			exists, _, _ := isAnnotatedForInjection(annotation)
+			Expect(exists).To(BeFalse())
```
Comment (michelleN) [inline pkg/injector/webhook_test.go:158]: Could we go ahead and also test the second and third return vars (enabled and err) are set to false and nil?

Comment (shashankram) [reply pkg/injector/webhook_test.go:158]: Done.

File: docs/patterns/sidecar_injection.md (hunk 1 above)
Comment (shashankram) [reply docs/patterns/sidecar_injection.md:19]: I added this is a prerequisite.

File: pkg/injector/webhook_test.go [hunk 4]
```diff
@@ -93,3 +97,290 @@ func (mc mockCertificate) GetCertificateChain() []byte           { return []byte
...
+				},
+			},
+			Spec: corev1.PodSpec{
+				ServiceAccountName: "test-SA",
+			},
+		}
+		_, err = fakeClientSet.CoreV1().Pods(namespace).Create(context.TODO(), podWithInjectAnnotationEnabled, metav1.CreateOptions{})
+		Expect(err).ToNot(HaveOccurred())
+
+		mockNsController.EXPECT().IsMonitoredNamespace(namespace).Return(true).Times(1)
```
Comment (michelleN) [inline pkg/injector/webhook_test.go:221]: obsessed. this is so cool!
